### PR TITLE
Consider removed runs when assigning points

### LINKED ISSUE #24: Consider removed runs when assigning points
One of the issues brought up when discussing the new points system was that runners could remove their runs from the leaderboard if they knew they'd lose points with their current placement. To prevent this, points should be calculated from the player's most recent logged submission (in `week.log`) regardless of if they've removed it. However, admins should still be able to remove invalid runs from the log completely.

Resolving this issue consists of performing three tasks:
1. Modifying the `points` util to consider _all_ submissions, ignoring removal markers;
2. Creating a `weeklog` command which cuts out a submission from the buffer entirely, without leaving a removal marker;
3. Modifying the `leaderboard` run removal command to allow for optionally purging the log entry.

### PR DESCRIPTION
**Pull request by p2r3**
Uses the weeklog to restore user-deleted runs when calculating elo deltas, and implements a way to distinguish user and admin run deletions. Tested on data copied from the live version at epochtal.p2r3.com. After accounting for edge cases, only 3 runs were considered "missing", each of which I checked individually to ensure that they were indeed removed by the player. Placement on the global leaderboard is hardly different, which I see as a good indication that everything is working correctly.

## Week 146 top 5 standings post-recalculation:
image

Edge cases specific to the epochtal.p2r3.com deployment are caused by faulty historical data. Patches for those will be applied to the deployment itself and are not part of this pull request. On a less related note, week logs would _really_ benefit from having an extra byte for flags to mark things like proof type, but that would require lots of awkward leaderboard scraping to grandfather that data into the existing logs. I simply do not have the time for this now. As such, the current implementation is forced to assume all deleted LP runs are segmented.

Closes #24

### CODE REVIEW

File: util/weeklog.js
Comment (PancakeTAS) [inline util/weeklog.js:32]: This seems incredibly scuffed.. what if the length is not one, but two characters short? Also, how did this piece of code work before this line as put in place?

File: util/points.js
Comment (PancakeTAS) [inline util/points.js:147]: I find this comment here to be a bit misleading as it implies the single line following it is the entire clone, rather than the entire following bit.

**Review by PancakeTAS — APPROVED** [review]
Approval, but do look at my two comments.
In general I find the weeklog being a binary file not actually that good of an idea. In my opinion the weeklog would be an amazing first module to port to a relational database if anyone wants to take a look at it.

File: util/weeklog.js
Comment (p2r3) [reply util/weeklog.js:32]: It's a while loop, so it'll leftpad until the length is 17.

It worked before because we've never attempted to use week logs with forged SteamIDs. The currently deployed system has some placeholder values for runners that came and went before we had Steam auth, and those IDs are basically a bunch of zeroes with an index at the end. The log used to parse this as a number, we need a string from it. I only now noticed that this is a problem, because there weren't many cases where you'd pass the SteamID from an ancient log into a util, until now.

Comment (PancakeTAS) [reply util/weeklog.js:32]: > It's a while loop

In my defense, I am eepy

**Comment by p2r3** [comment]
Completely agreed, I also dislike how hacky the logs have become. Definitely an excellent application for a more coherent database, though I'm not currently in the position to implement that.

**Comment by soni801** [comment]
I'd really like to work on db integration, but my understanding was that you didn't really want to move over to that just yet? I can absolutely write the implementation myself (so no time spent for you), but I'm waiting on your seal of approval to go ahead with the dev effort, aint gonna do it if its just gonna get rejected anyway lmao

**Comment by PancakeTAS** [comment]
As far as I know p2r3 is studying databases right these few days, so perhaps you should wait up until that class is done?

File: util/weeklog.js
Comment (soni801) [inline util/weeklog.js:56]: I find this to be pretty hard to read, but I also don't know this code too much, so as long as there's a good reason why you're doing it this way I guess I won't complain

File: util/weeklog.js
Comment (soni801) [inline util/weeklog.js:70]: This refactor makes no sense? Isn't the previous version much simpler to read here?

**Review by soni801 — APPROVED** [review]
I don't really know that this would be easy to test, so I'm gonna trust that your test was good enough :D

Code seems good, only some very minor comments this time.

**Comment by p2r3** [comment]
> I'd really like to work on db integration, but my understanding was that you didn't really want to move over to that just yet?

I don't think a full db integration system-wide would be a good idea right now, both because I wouldn't be competent to review that, and because many still growing parts of the system would be impacted. Many utils still confidently assume the `context.[data|file].[week|leaderboard|log|...]` structure, and it's best to not touch that yet.

_However_, I see no downside to rewriting compact week logs to use a relational database. The scope of that wouldn't be extensive, there's no util that directly tries to parse the binary format without going through the `weeklog` util first, so it's already abstracted, and I think I also have enough experience now to review a relatively small-scale change like this. Baby steps.

**Pull request merged**
